File: ingestion/json/buildingFolder.py
from datetime import date
import json
import os
from pathlib import Path
import shutil

import pandas as pd
from pangres import upsert
from sqlalchemy import create_engine, text

def main():
    # Define directories matching your local paths
    input_folder = Path(
        os.getenv(
            "INPUT_FOLDER",
            "/Users/jonathankee/Data-Science-Projects/ingestion/sources_unprocessed",
        )
    )
    processed_folder = Path(
        "/Users/jonathankee/Data-Science-Projects/ingestion/sources_processed"
    )

    if not input_folder.exists():
        print(f"Input directory does not exist: {input_folder}")
        return

    # Filter files: must end with '.json' and NOT contain 'AI1'
    target_files = []
    for f in input_folder.iterdir():
        if f.is_file() and f.name.endswith(".json") and "AI1" not in f.name:
            target_files.append(f)

    if len(target_files) == 0:
        print('No eligible files (without "AI1") found in the directory.')
        return

    buildings_data = []
    building_costs_data = []
    current_date = date.today().isoformat()

    # Parse JSON files
    for file_path in target_files:
        try:
            with open(file_path, "r", encoding="utf-8") as f:
                data_payload = json.load(f)
            print(f"Loaded JSON successfully from {file_path.name}!")
        except Exception as e:
            print(f"Failed to read or parse JSON file {file_path.name}: {e}")
            continue

        building_id = data_payload.get("BuildingId")
        if not building_id:
            print(f'Missing "BuildingId" in JSON payload for file {file_path.name}.')
            continue

        buildings_data.append({
            "id": building_id,
            "ticker": data_payload.get("Ticker"),
            "name": data_payload.get("Name"),
            "area_cost": data_payload.get("AreaCost"),
            "user_name_submitted": data_payload.get("UserNameSubmitted"),
            "file_date": current_date,
        })

        for cost in data_payload.get("BuildingCosts", []):
            building_costs_data.append({
                "building_id": building_id,
                "commodity_name": cost.get("CommodityName"),
                "commodity_ticker": cost.get("CommodityTicker"),
                "weight": cost.get("Weight"),
                "volume": cost.get("Volume"),
                "amount": cost.get("Amount"),
            })

    if len(buildings_data) == 0:
        print("No valid building data found to process.")
        return

    # Convert to Pandas DataFrames and deduplicate to ensure within-batch idempotency
    df_buildings = pd.DataFrame(buildings_data).drop_duplicates(subset=["id"], keep="last")
    df_costs = pd.DataFrame(building_costs_data).drop_duplicates(subset=["building_id", "commodity_ticker"], keep="last")

    # pangres requires primary keys to be set as the DataFrame index
    df_buildings = df_buildings.set_index("id")
    df_costs = df_costs.set_index(["building_id", "commodity_ticker"])

    # Database connection using SQLAlchemy
    engine = create_engine("postgresql://postgres:abc123@localhost:5432/prosperous_universe")

    try:
        with engine.begin() as conn:
            # Ensure schema exists
            conn.execute(text("CREATE SCHEMA IF NOT EXISTS raw;"))

        # ---------------------------------------------------------
        # Upsert for 'buildings' Table via pangres
        # ---------------------------------------------------------
        upsert(
            con=engine,
            df=df_buildings,
            table_name="buildings",
            schema="raw",
            if_row_exists="update",
            add_new_columns=True
        )

        # ---------------------------------------------------------
        # Upsert for 'building_costs' Table via pangres
        # ---------------------------------------------------------
        if not df_costs.empty:
            upsert(
                con=engine,
                df=df_costs,
                table_name="building_costs",
                schema="raw",
                if_row_exists="update",
                add_new_columns=True
            )

        print('\nSUCCESS: Pipeline execution completed with pangres Upserts in schema "raw".\n')

        # Safe check for database results
        try:
            query = """
                SELECT id, name, ticker, area_cost, user_name_submitted, file_date 
                FROM raw.buildings
            """
            with engine.connect() as conn:
                summary_df = pd.read_sql(text(query), conn)
                
            print('--- Queried Database Summary (from schema "raw") ---')
            for row in summary_df.to_dict(orient="records"):
                print(row)
        except Exception as query_err:
            print(f"Note: Could not run summary print query automatically: {query_err}")

        # Move processed files to the processed folder
        processed_folder.mkdir(parents=True, exist_ok=True)
        for file_path in target_files:
            processed_file_path = processed_folder / file_path.name
            shutil.move(str(file_path), str(processed_file_path))
            print(f"Successfully moved JSON file to: {processed_file_path}")

    except Exception as e:
        print(f"Error during pipeline execution: {e}\n")


if __name__ == "__main__":
    main()

File: ingestion/json/buildingFolder.ts
import { readFile, mkdir, rename, readdir } from 'node:fs/promises';
import { join, basename } from 'node:path';
import { Sequelize, DataTypes, Model } from 'sequelize';

// 2. Initialize Sequelize with the 'raw' schema
const sequelize = new Sequelize('prosperous_universe', 'postgres', 'abc123', {
  host: 'localhost',
  dialect: 'postgres',
  logging: false,
  schema: 'raw',
});

// 3. Define the Building Model
class Building extends Model {
  declare id: string;
  declare ticker: string;
  declare name: string;
  declare areaCost: number;
  declare submittedBy: string;
}

Building.init(
  {
    id: {
      type: DataTypes.STRING,
      primaryKey: true,
    },
    ticker: {
      type: DataTypes.STRING,
      allowNull: true, // Made flexible to prevent validation errors if missing
    },
    name: {
      type: DataTypes.STRING,
      allowNull: true,
    },
    areaCost: {
      type: DataTypes.INTEGER,
      field: 'area_cost',
      allowNull: true,
    },
    submittedBy: {
      type: DataTypes.STRING,
      field: 'user_name_submitted',
      allowNull: true,
    },
  },
  {
    sequelize,
    modelName: 'Building',
    tableName: 'buildings',
    schema: 'raw'
  }
);

// 4. Define the BuildingCost Model
class BuildingCost extends Model {
  declare id: number;
  declare commodityName: string;
  declare commodityTicker: string;
  declare weight: number;
  declare volume: number;
  declare amount: number;
  declare buildingId: string;
}

BuildingCost.init(
  {
    id: {
      type: DataTypes.INTEGER,
      autoIncrement: true,
      primaryKey: true,
    },
    commodityName: {
      type: DataTypes.STRING,
      allowNull: true,
      field: 'commodity_name',
    },
    commodityTicker: {
      type: DataTypes.STRING,
      allowNull: true,
      field: 'commodity_ticker',
    },
    weight: DataTypes.FLOAT,
    volume: DataTypes.FLOAT,
    amount: {
      type: DataTypes.INTEGER,
      allowNull: true,
    },
  },
  {
    sequelize,
    modelName: 'BuildingCost',
    tableName: 'building_costs',
    schema: 'raw'
  }
);

// 5. Define Relational Associations
Building.hasMany(BuildingCost, {
  foreignKey: 'buildingId',
  as: 'buildingCosts',
  onDelete: 'CASCADE',
});
BuildingCost.belongsTo(Building, {
  foreignKey: 'buildingId',
});

// 6. Data Seeding Function with Transaction Handling
async function seedData(dataPayload: any) {
  const buildingId = dataPayload.BuildingId;
  
  if (!buildingId) {
    throw new Error('Missing "BuildingId" in JSON payload.');
  }

  const buildingData = {
    id: buildingId,
    ticker: dataPayload.Ticker || null,
    name: dataPayload.Name || null,
    areaCost: dataPayload.AreaCost ?? null,
    submittedBy: dataPayload.UserNameSubmitted || null,
  };

  const buildingCostsData = (dataPayload.BuildingCosts || []).map((cost: any) => ({
    commodityName: cost.CommodityName || null,
    commodityTicker: cost.CommodityTicker || null,
    weight: cost.Weight ?? null,
    volume: cost.Volume ?? null,
    amount: cost.Amount ?? null,
    buildingId: buildingId,
  }));

  // Use a transaction to safely handle existing data updates
  const t = await sequelize.transaction();

  try {
    // Upsert the main building record
    await Building.upsert(buildingData, { transaction: t });

    // Clear old dependent costs for this building to prevent duplicates/conflicts
    await BuildingCost.destroy({ where: { buildingId }, transaction: t });

    // Insert new costs if any exist
    if (buildingCostsData.length > 0) {
      await BuildingCost.bulkCreate(buildingCostsData, { transaction: t });
    }

    await t.commit();
    console.log(`\nSUCCESS: Upserted Building "${buildingData.name || buildingId}" into schema "raw".\n`);
  } catch (error) {
    await t.rollback();
    throw error;
  }
}

// 7. Query Function
async function getBuilding(buildingId: string) {
  const result = await Building.findByPk(buildingId, {
    include: [
      {
        model: BuildingCost,
        as: 'buildingCosts',
      },
    ],
  });

  console.log('--- Queried Database Result (from schema "raw") ---');
  console.log(JSON.stringify(result, null, 2));
}

async function readJson(filePath: any) {
  try {
    const data = await readFile(filePath, 'utf-8');
    const json = JSON.parse(data);

    console.log(`Loaded JSON successfully from ${basename(filePath)}!`);
    return json;
  } catch (error: any) {
    console.error(`Failed to read or parse JSON file ${basename(filePath)}:`, error.message);
  }
}

// 8. Main Execution Wrapper
async function main() {
  const inputFolder = process.argv[2] || '/Users/jonathankee/Data-Science-Projects/ingestion/sources_unprocessed';
  const processedFolder = '/Users/jonathankee/Data-Science-Projects/ingestion/sources_processed';

  try {
    await sequelize.authenticate();
    console.log('PostgreSQL connection established successfully.');

    await sequelize.createSchema('raw', {}).catch(() => {});
    await sequelize.sync({ force: false });

    const files = await readdir(inputFolder);
    const targetFiles = files.filter(file => file.endsWith('.json') && !file.endsWith('AI1.json'));

    if (targetFiles.length === 0) {
      console.log('No eligible files (not ending with AI1.json) found in the directory.');
      return;
    }

    for (const file of targetFiles) {
      const jsonFilePath = join(inputFolder, file);

      let data = await readJson(jsonFilePath);
      if (!data) continue;

      try {
        await seedData(data);

        if (data.BuildingId) {
          await getBuilding(data.BuildingId);
        }

        const processedFilePath = join(processedFolder, file);
        await mkdir(processedFolder, { recursive: true });
        await rename(jsonFilePath, processedFilePath);
        console.log(`Successfully moved JSON file to: ${processedFilePath}`);
      } catch (error: any) {
        console.error(`Error processing file ${file}:`, error.message);
        if (error.errors && Array.isArray(error.errors)) {
          error.errors.forEach((err: any) => {
            console.error(`   -> Validation Detail [${err.path}]: ${err.message}`);
          });
        }
      }
    }
  } catch (error) {
    console.error('Database Operation Failed:', error);
  } finally {
    await sequelize.close();
  }
}

main();